sdk/core/AzureCore/Source/DataStructures/Iso8601Duration.swift

```diff
@@ -0,0 +1,200 @@
+// --------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// The MIT License (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the ""Software""), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+import Foundation
+
+private func components(forString string: String, designatorSet: CharacterSet) -> [String: String] {
+    if string.isEmpty {
+        return [:]
+    }
+
+    var decimalSet = CharacterSet(charactersIn: ".,")
+    decimalSet.formUnion(.decimalDigits)
+    let componentValues = string.components(separatedBy: designatorSet).filter { !$0.isEmpty }
+    let designatorValues = string.components(separatedBy: decimalSet).filter { !$0.isEmpty }
+
+    guard componentValues.count == designatorValues.count else {
+        return [:]
+    }
+    return Dictionary(uniqueKeysWithValues: zip(designatorValues, componentValues))
+}
+
+private func parse(components: [String: String], forKey key: String) -> (whole: Int?, decimal: Double?)? {
+    guard let parsed = components[key],
+        let double = Double(parsed.replacingOccurrences(of: ",", with: ".")) else { return nil }
+    let whole = Int(double)
+    let decimal = (double - Double(whole))
```

Comment:
This casting is a little hard to follow. But I can't really think of a better way to make this cleaner. I am assuming it is always safe to round down?